src/software/ai/hl/stp/action/chip_action.cpp

```diff
@@ -67,61 +67,69 @@ void ChipAction::calculateNextIntent(IntentCoroutine::push_type& yield)
-    }
-    else
-    {
-        yield(std::make_unique<ChipIntent>(robot->id(), chip_origin, chip_direction,
-                                           chip_distance_meters, 0));
-    }
+        // ASCII art showing the region behind the ball
+        // Diagram not to scale
+        //
+        //                             X
+        //                      v-------------v
+        //
+        //                   >  B-------------C
+        //                   |   \           /
+        //                   |    \         /
+        //                   |     \       /       <- Region considered "behind ball"
+        //                 X |      \     /
+        //                   |       \   /
+        //                   |        \ /
+        //    The ball is    >         A
+        //    at A
+        //                             |
+        //                             V
+        //                     direction of chip
+
+        // A vector in the direction opposite the chip (behind the ball)
+
+        Vector behind_ball =
+            Vector::createFromAngle(this->chip_direction + Angle::half());
+
+
+        // The points below make up the triangle that defines the region we treat as
+        // "behind the ball". They correspond to the vertices labeled 'A', 'B', and 'C'
+        // in the ASCII diagram
+
+        // We make the region close enough to the ball so that the robot will still be
+        // inside it when taking the chip.
+        Point behind_ball_vertex_A = chip_origin;
+        Point behind_ball_vertex_B =
+            behind_ball_vertex_A + behind_ball.normalize(size_of_region_behind_ball) +
+            behind_ball.perpendicular().normalize(size_of_region_behind_ball / 2);
+        Point behind_ball_vertex_C =
+            behind_ball_vertex_A + behind_ball.normalize(size_of_region_behind_ball) -
+            behind_ball.perpendicular().normalize(size_of_region_behind_ball / 2);
+
+        Triangle behind_ball_region =
+            Triangle(behind_ball_vertex_A, behind_ball_vertex_B, behind_ball_vertex_C);
+
+        bool robot_behind_ball = behind_ball_region.contains(robot->position());
+        // The point in the middle of the region behind the ball
+        Point point_behind_ball =
+            chip_origin + behind_ball.normalize(size_of_region_behind_ball * 3 / 4);
+
+        // If we're not in position to chip, move into position
+        if (!robot_behind_ball)
+        {
+            yield(std::make_unique<MoveIntent>(
+                robot->id(), point_behind_ball, chip_direction, 0.0, 0,
+                DribblerEnable::OFF, MoveType::NORMAL, AutokickType::NONE,
+                BallCollisionType::ALLOW));
+        }
+        else
+        {
+            yield(std::make_unique<ChipIntent>(robot->id(), chip_origin, chip_direction,
+                                               chip_distance_meters, 0));
+            break;
+        }
+
+    } while (true);
```

Comment:
We shouldn't have `while true` here since this will loop forever which doesn't obey the `loop_forever=false`. The "done" condition should go here. I don't think we have one properly made yet for kicking/chipping, so can you make a ticket to implement one and use it here?

Reply:
I think it'd make more sense to implement the done condition before making this change, because of what Mathew said

Reply:
The `break` statement after the chipIntent is yielded (line 131) was meant to signify reaching a done state, since after we break from the `while(true)` loop the coroutine ends and thus the action is done. 

So Im not 100% sure if by "done condition" you mean making it clearer (like a local bool `has_chip_intent_yielded` which gets set to true once we yield the chipIntent) or something else that Im missing?

Reply:
Ah I missed the `break` statement.

By "done condition", this is usually the condition(s) that are synonymous with the Action being "done". For example, a `StopAction` is done once the velocity of the robot is below some threshold. Likewise, the done condition for a `ChipAction` is when the ball has been chipped (not just when we ask it to chip, when the action actually takes place IRL).

Right now, simply `yielding` the ChipIntent a single time isn't quite the same as being done, because we don't know that the robot chipped the ball successfully IRL yet, and that's what we're waiting for.

To avoid the `break` statement, I think we can write the action this way to make better use of coroutines:
```
do {
  // yield Move behind ball
}while(!robot_behind_ball);

do {
  // yield Chip Intent
}while(!ball has not been chipped);
```

I agree with @jonathanlew that we should probably implement a function to detect when a chip/kick has taken place, so we can use it as the "done condition" in this PR to avoid breaking expected behavior. I'll make a ticket.

Reply:
Ok I've created #1338 . @amrm3lm do you mind quickly doing that first?

Reply:
I'll add that this would be a good situation to have a test case to catch if we're breaking expected behaviour

Reply:
> because we don't know that the robot chipped the ball successfully IRL yet, and that's what we're waiting for.

Oh that makes a lot of sense, thanks for clarifying. I'll do #1338 and add tests before coming back to this. 

One question though, what do we expect to happen if we call `update_control_parameters` in either of the phases? A reason the `while(true)` loop + `break` made sense to me was that if we updated the params while attempting to chip/kick, then we'd be able to go back and reevaluate whether we're behind the ball. Or do we also expect tactics to complete the action before making significant changes to it.